src/backend/api/meals.js

```diff
@@ -1,14 +1,203 @@
-const express = require("express");
+const express = require('express');
 const router = express.Router();
-const knex = require("../database");
+const knex = require('../database');
+
+router.get('/', async (req, res) => {
+  let query = knex.select('meal.*').from('meal');
+
+  if ('maxPrice' in req.query) {
+    const maxPrice = parseInt(req.query.maxPrice);
+
+    if (!maxPrice) {
+      res.status(404).json({ error: 'You did not enter max price' });
+    } else {
+      meals = query.where('price', '<', maxPrice);
+    }
+  }
+
+  if ('availableReservations' in req.query) {
+    const availableReservations = req.query.availableReservations;
+
+    meals = query
+      .join('reservation', 'reservation.meal_id', '=', 'meal.id')
+      .groupBy('meal.id');
+    if (availableReservations === 'true') {
+      meals = meals.where(
+        'meal.max_reservations',
+        '>',
+        'SUM(reservation.number_of_guests)'
+      );
+    } else {
+      meals = meals.where(
+        'meal.max_reservations',
+        '=',
+        'SUM(reservation.number_of_guests)'
+      );
+    }
+  }
+
+  if ('title' in req.query) {
+    const title = String(req.query.title);
+
+    if (!title) {
+      res.status(404).json({ error: 'You did not enter title' });
+    } else {
+      meals = query.where('title', 'like', `%${title}%`);
+    }
+  }
+
+  if ('dateAfter' in req.query) {
+    const dateAfter = new Date(req.query.dateAfter);
+
+    if (!dateAfter) {
+      res.status(404).json({ error: 'You did not enter date' });
+    } else {
+      meals = query.where('when', '>', dateAfter);
+    }
+  }
+
+  if ('dateBefore' in req.query) {
+    const dateBefore = new Date(req.query.dateAfter);
+
+    if (!dateBefore) {
+      res.status(404).json({ error: 'You did not enter date' });
+    } else {
+      meals = query.where('when', '<', dateBefore);
+    }
+  }
+
+  if ('limit' in req.query) {
+    const limit = parseInt(req.query.limit);
+
+    if (!limit) {
+      res.status(404).json({ error: 'You did not enter limit' });
+    } else {
+      meals = query.limit(limit);
```

Comment:
instead of using the query, you could use the meals through all if statements, so that all the queries apply to the same array to filter.